### PR TITLE
Feature/corona tests

### PR DESCRIPTION
**Pull request by davidkydd**
Runner test for Corona - requires corona appPackage to already exist on templateRunner batch account (it does), also adds a new argument to the runner cmd line - keyvaultUrl - which is added after storageKey and before ServicePrincipal args. SP needs get access to the keyvault.

NOTE on merge to master, pipeline needs to be updated to include the new keyvaultUrl argument for  runner.py

### CODE REVIEW

File: runner/runner.py [hunk 1]
```diff
@@ -45,6 +46,19 @@ def create_batch_client(args: object) -> batch.BatchExtensionsClient:
         base_url=args.BatchAccountUrl,
         subscription_id=args.BatchAccountSub)
 
+def create_keyvault_client(args: object):
+    """
+    Create keyvault client namedTuple with url
+    """
```
Comment (matthew-rollitt) [inline runner/runner.py:56]: Can you add the correct param, rtype and return docs to this and since it returns a type you need to add a ->  #Resolved

File: runner/runner.py [hunk 2]
```diff
@@ -64,6 +78,7 @@ def runner_arguments():
     parser.add_argument("BatchAccountSub", help="The batch account sub ")
     parser.add_argument("StorageAccountName", help="Storage name ")
     parser.add_argument("StorageAccountKey", help="storage key")
+    parser.add_argument("KeyVaultUrl", help="Azure Key vault to fetch secrets from, service principal must have access")
     parser.add_argument("ServicePrincipalCredentialsClientID", help="Service Principal id")
```
Comment (matthew-rollitt) [inline runner/runner.py]: if you make it an argument like so you'll always need to always add this to your command line. I would recommend making this optional since you can use this runner without the corona renderer and this arg wouldn't be needed.

File: runner/runner.pyproj [hunk 3]
```diff
@@ -0,0 +1,39 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" DefaultTargets="Build">
+  <PropertyGroup>
```
Comment (matthew-rollitt) [inline runner/runner.pyproj:3]: Do you have to add these two VS files?

File: runner/runner.sln [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 15
+VisualStudioVersion = 15.0.28307.539
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{888888A0-9F3D-457C-B088-3A5042F75D52}") = "runner", "runner.pyproj", "{1BBC824F-E02F-4E05-AF63-615CDA90775D}"
```
Comment (matthew-rollitt) [inline runner/runner.sln:6]: why is this file added?

Comment (davidkydd) [reply runner/runner.sln:6]: As in comment for commit, so project can be opened in visual studio. Much easier to manage python environments, install requirements etc

File: runner/job_manager.py [hunk 5]
```diff
@@ -413,3 +419,12 @@ def delete_resources(self, batch_service_client: batch.BatchExtensionsClient,
             logger.info(
                 "Job: {}. did not complete successfully, Container {} was not deleted.".format(
                     self.job_id, self.storage_info.output_container))
+
+    def update_params_with_values_from_keyvault(self, parameters, keyvault_client_with_url):
```
Comment (matthew-rollitt) [inline runner/job_manager.py]: >update_params_with_values_from_keyvault [](start = 8, length = 39)

This needs all the docs rtype/return/params ':Type' and -> type 

Can you move this to the Utils.py file. That file handles the utility functions that job_manager shouldn't handle.

File: runner/job_manager.py [hunk 6]
```diff
@@ -413,3 +419,12 @@ def delete_resources(self, batch_service_client: batch.BatchExtensionsClient,
...
+        for parameter, value in parameters.items():
+            stringValue = str(value)[11:-2] #substring the inner value from format {"value": "<value-we-want>"}
+            if len(stringValue) > 0 and stringValue.startswith(_keyVault_param_identifier):
+                secretId = stringValue[len(_keyVault_param_identifier):]  #substring the original value without the keyvault identifier prefix
+                secret_bundle = keyvault_client_with_url[0].get_secret(keyvault_client_with_url[1], secretId, '')
+                parameters[parameter] = secret_bundle.value
```
Comment (matthew-rollitt) [inline runner/job_manager.py]: Can you also use pep8, it's similar to style cop

File: runner/runner.pyproj (hunk 3 above)
Comment (davidkydd) [reply runner/runner.pyproj:3]: If you don't want to open as VS project then just don't and no impact. I have found this much more productive way to manage python projects

File: runner/job_manager.py [hunk 7]
```diff
@@ -413,3 +419,12 @@ def delete_resources(self, batch_service_client: batch.BatchExtensionsClient,
             logger.info(
                 "Job: {}. did not complete successfully, Container {} was not deleted.".format(
                     self.job_id, self.storage_info.output_container))
+
+    def update_params_with_values_from_keyvault(self, parameters, keyvault_client_with_url):
+        for parameter, value in parameters.items():
+            stringValue = str(value)[11:-2] #substring the inner value from format {"value": "<value-we-want>"}
```
Comment (matthew-rollitt) [inline runner/job_manager.py]: >stringValue [](start = 12, length = 11)

string_value... also may want a different name

File: runner/runner.py (hunk 2 above)
Comment (davidkydd) [reply runner/runner.py]: Can do, while only used for Corona now, keyVault params are generally available to other templates which might want to use the capability

File: runner/job_manager.py [hunk 8]
```diff
@@ -413,3 +419,12 @@ def delete_resources(self, batch_service_client: batch.BatchExtensionsClient,
...
+            if len(stringValue) > 0 and stringValue.startswith(_keyVault_param_identifier):
+                secretId = stringValue[len(_keyVault_param_identifier):]  #substring the original value without the keyvault identifier prefix
```
Comment (matthew-rollitt) [inline runner/job_manager.py]: >secretId [](start = 16, length = 8)

secret_Id like secret_bundle

File: runner/job_manager.py [hunk 9]
```diff
@@ -15,6 +17,7 @@
 This module is responsible for creating, submitting and monitoring the pools and jobs
 
 """
+_keyVault_param_identifier = "KEYVAULT_"
```
Comment (matthew-rollitt) [inline runner/job_manager.py]: Can you move this to the utils.py file like the method below?

File: azure-pipelines.yml [hunk 10]
```diff
@@ -34,7 +34,7 @@ steps:
     . .azure-pipelines/exec.ps1
     $ErrorActionPreference = "Stop"
     pushd '$(projectRoot)'
-    exec { python runner.py "Tests/TestConfiguration.json" "$(PS_BATCH_ACCOUNT_NAME)" "$(runnerBatchAccountSecret)" "$(PS_BATCH_ACCOUNT_URL)" "$(PS_BATCH_ACCOUNT_SUB)" "$(PS_STORAGE_ACCOUNT_NAME)" "$(StorageAccountKey)" "$(PS_SERVICE_PRINCIPAL_CREDENTIALS_CLIENT_ID)" "$(ServicePrincipalCredentialsSecret)" "$(PS_SERVICE_PRINCIPAL_CREDENTIALS_TENANT)" "$(PS_SERVICE_PRINCIPAL_CREDENTIALS_RESOUCE)"}
+    exec { python runner.py "Tests/TestConfiguration.json" "$(PS_BATCH_ACCOUNT_NAME)" "$(runnerBatchAccountSecret)" "$(PS_BATCH_ACCOUNT_URL)" "$(PS_BATCH_ACCOUNT_SUB)" "$(PS_STORAGE_ACCOUNT_NAME)" "$(StorageAccountKey)" "$(PS_KEYVAULT_URL)" "$(PS_SERVICE_PRINCIPAL_CREDENTIALS_CLIENT_ID)" "$(ServicePrincipalCredentialsSecret)" "$(PS_SERVICE_PRINCIPAL_CREDENTIALS_TENANT)" "$(PS_SERVICE_PRINCIPAL_CREDENTIALS_RESOUCE)"}
```
Comment (matthew-rollitt) [inline azure-pipelines.yml]: >"$(PS_KEYVAULT_URL)" [](start = 220, length = 20)

can you make this optional?

File: runner/job_manager.py [hunk 11]
```diff
@@ -1,4 +1,6 @@
 from azure.storage.blob.models import ContainerPermissions
+from azure.keyvault import KeyVaultClient, KeyVaultAuthentication
+from azure.common.credentials import ServicePrincipalCredentials
```
Comment (matthew-rollitt) [inline runner/job_manager.py]: Can this be moved to utils.py?

File: runner/runner.sln (hunk 4 above)
Comment (matthew-rollitt) [reply runner/runner.sln:6]: Fair enough, just use pycharm since that gives you all the pep8 stuff

File: runner/job_manager.py (hunk 7 above)
Comment (davidkydd) [reply runner/job_manager.py]: Thanks - inner_value is clearer

File: runner/job_manager.py (hunk 6 above)
Comment (davidkydd) [reply runner/job_manager.py]: PyLint gives 86 flags, only a couple of line-too-longs in my code...

File: azure-pipelines.yml (hunk 10 above)
Comment (davidkydd) [reply azure-pipelines.yml]: Done

File: runner/job_manager.py (hunk 11 above)
Comment (davidkydd) [reply runner/job_manager.py]: Done

File: runner/job_manager.py (hunk 8 above)
Comment (davidkydd) [reply runner/job_manager.py]: Done

File: runner/job_manager.py (hunk 5 above)
Comment (davidkydd) [reply runner/job_manager.py]: Done

File: runner/runner.py (hunk 1 above)
Comment (davidkydd) [reply runner/runner.py:56]: Done

File: runner/job_manager.py (hunk 9 above)
Comment (davidkydd) [reply runner/job_manager.py]: Done

File: runner/job_manager.py [hunk 12]
```diff
@@ -90,6 +91,7 @@ def create_and_submit_job(self, batch_client: batch.BatchExtensionsClient):
         template = ctm.load_file(self.template_file)
         parameters = ctm.load_file(self.parameters_file)
 
+        utils.update_params_with_values_from_keyvault(parameters, self.keyvault_client_with_url)
         # overrides some of the parameters needed in the file, container SAS
```
Comment (matthew-rollitt) [inline runner/job_manager.py:96]: Can you add a small # comment above this saying what this is needed for?

Comment (davidkydd) [reply runner/job_manager.py:96]: Believe its pretty self explanatory but will just for you

File: runner/utils.py [hunk 13]
```diff
@@ -283,3 +284,32 @@ def execute_parallel_jobmanagers(method_name: str, job_managers: 'list[job_manag
     # Wait for all threads to finish
     for thread in threads:
         thread.join()
+
+def update_params_with_values_from_keyvault(parameters: dict(), keyvault_client_with_url: tuple()):
+    """
+    Replaces parameters with keyvault placeholder values with the actual value fetched from keyvault
+
```
Comment (matthew-rollitt) [inline runner/utils.py:292]: Maybe you can add a comment about this being used for corona renderer in here. Since this is only used for the corona renderer

File: runner/runner.py [hunk 14]
```diff
@@ -1,5 +1,6 @@
 from __future__ import print_function
 from azure.common.credentials import ServicePrincipalCredentials
+from azure.keyvault import KeyVaultClient
 import traceback
```
Comment (matthew-rollitt) [inline runner/runner.py:4]: What do you think about making this an 'as keyVault' import so all the calls go something like this keyVault.KeyVaultClient( ?

Comment (davidkydd) [reply runner/runner.py:4]: Due to the size of the project we are unlikely to have any namespace collisions with KeyVaultClient, so think as it is currently is fine. If it was called "Client" we might want to import it behind a namespace prefix

File: runner/utils.py (hunk 13 above)
Comment (davidkydd) [reply runner/utils.py:292]: Currently only used for corona renderer, but it is a general solution for passing secrets as template parameter values. Have added an e.g. for the Corona case

**Review by matthew-rollitt — APPROVED** [review]
:shipit:
